compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -596,11 +616,15 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
         }
     }
 
-    // If a function uses #[target_feature] it can't be inlined into general
+    // If a function uses `#[target_feature]` it can't be inlined into general
     // purpose functions as they wouldn't have the right target features
-    // enabled. For that reason we also forbid #[inline(always)] as it can't be
+    // enabled. For that reason we also forbid `#[inline(always)]` as it can't be
     // respected.
-    if !codegen_fn_attrs.target_features.is_empty() && codegen_fn_attrs.inline == InlineAttr::Always
+    //
+    // `#[rustc_force_inline]` doesn't need to be prohibited here, that
+    // is implemented entirely in rustc can attempt to inline and error if it cannot.
+    if !codegen_fn_attrs.target_features.is_empty()
+        && matches!(codegen_fn_attrs.inline, InlineAttr::Always)
```

Comment:
If I checked for `codegen_fn_attrs.inline.always()` here then I'd also prohibit `target_feature`s from being used with `rustc_force_inline`, which I don't believe is necessary.

We prohibit `target_features` and `inline(always)` because of [blockers w/ LLVM](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/inline.28always.29.20.2B.20target_feature/near/481124667), but that doesn't apply to forced inlining because that all happens in rustc.

We already have checks for target features in the MIR inliner and can rely on those for forced inlining.

Reply:
I see, thanks for explaining.